File: LC437.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def psum(self, root, x):
        if not root:
            return 0

        return (x == root.val) + self.psum(root.left, x - root.val) + self.psum(root.right, x - root.val)
    
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root:
            return 0
        
        return self.psum(root, targetSum) + self.pathSum(root.left, targetSum) + self.pathSum(root.right, targetSum)

File: LC437.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {

public:
    int count = 0;
    int psum(TreeNode* root, long long x){
        if(!root)
            return 0;
        
        return (x == root->val) + psum(root->left, x - root->val) + psum(root->right, x - root->val);
    }

    int pathSum(TreeNode* root, int target) {
        
        if(!root)
            return 0;

        return psum(root, (long long) target) + pathSum(root->left, (long long)target) + pathSum(root->right, (long long)target);

    }
};
